support-frontend/assets/pages/[countryGroupId]/components/formOnSubmit.ts

```diff
@@ -0,0 +1,204 @@
+	contributionAmount: number | undefined;
+}) => {
+	const personalData = extractPersonalDataFromForm(formData);
+	const { billingAddress, deliveryAddress } = hasDeliveryAddress
+		? extractDeliverableAddressDataFromForm(formData)
+		: extractNonDeliverableAddressDataFromForm(formData);
+
+	const ophanIds = getOphanIds();
+	const referrerAcquisitionData = {
+		...getReferrerAcquisitionData(),
+		labels: ['generic-checkout'], // Shall we get rid of this now?
+	};
+
+	const firstDeliveryDate =
+		productKey === 'TierThree'
+			? formatMachineDate(getTierThreeDeliveryDate())
+			: null;
+
+	const promoCode = promotion?.promoCode;
+	const appliedPromotion =
+		promoCode !== undefined
+			? {
+					promoCode,
+					countryGroupId: geoId,
+			  }
+			: undefined;
+	const supportAbTests = getSupportAbTests(abParticipations);
+
+	const createSupportWorkersRequest: RegularPaymentRequest = {
+		...personalData,
+		billingAddress,
+		deliveryAddress,
+		firstDeliveryDate,
+		paymentFields,
+		appliedPromotion,
+		ophanIds,
+		referrerAcquisitionData,
+		product: productFields,
+		supportAbTests,
+		debugInfo: '',
+	};
+
+	const createResponse = await createNewSupportWorkersExecution(
+		createSupportWorkersRequest,
+	);
+	const { userType, processPaymentResponse } = await pollForStatusOfExecution(
+		createResponse,
+	);
+
+	if (
+		processPaymentResponse.status === 'success' ||
+		processPaymentResponse.status === 'pending'
+	) {
+		goToThankYouPage(
+			productKey,
+			ratePlanKey,
+			promoCode,
+			userType,
+			contributionAmount,
+			personalData,
+			paymentMethod,
+			processPaymentResponse.status,
+			geoId,
+		);
+	} else {
+		console.error(
+			'processPaymentResponse error:',
+			processPaymentResponse.failureReason,
+		);
+		throw new FormSubmissionError(
+			'Sorry, something went wrong.',
+			appropriateErrorMessage(
+				processPaymentResponse.failureReason ?? 'unknown',
+			),
+		);
+	}
+};
+
+const createNewSupportWorkersExecution = async (
+	createSupportWorkersRequest: RegularPaymentRequest,
+): Promise<Response> => {
+	return fetch('/subscribe/create', {
+		method: 'POST',
+		headers: {
+			'Content-Type': 'application/json',
+		},
+		body: JSON.stringify(createSupportWorkersRequest),
+	});
+};
+
+const pollForStatusOfExecution = async (
+	createResponse: Response,
+): Promise<{
+	userType: UserType | undefined;
+	processPaymentResponse: ProcessPaymentResponse;
+}> => {
+	if (createResponse.ok) {
+		const statusResponse =
+			(await createResponse.json()) as CreateSubscriptionResponse;
+		const processPaymentResponse = await processPaymentWithRetries(
+			statusResponse,
+		);
+		return {
+			userType: statusResponse.userType,
+			processPaymentResponse,
+		};
+	} else {
+		const errorReason = (await createResponse.text()) as ErrorReason;
+		return {
+			userType: undefined,
+			processPaymentResponse: {
+				status: 'failure',
+				failureReason: errorReason,
+			},
+		};
+	}
+};
+
+const goToThankYouPage = (
```

Comment:
Not necessarily for now, but perhaps in future the return value of `submitForm` could be used to represent success or failure to the caller? The `success` status could even include the URL to redirect to, e.g.

```
{
  status: "success",
  redirectPath: "/uk/thank-you..."
}
```

Then we'd decouple this code from needing to know about the `window`.